src/Core/OrganizationFeatures/AuthorizationHandlers/CollectionUserAuthorizationHandler.cs

```diff
@@ -0,0 +1,207 @@
+            if (targetCollectionsForUser.Any(tc => tc.OrganizationId != targetUser.OrganizationId))
+            {
+                context.Fail();
+                return;
+            }
+        }
+
+        context.Succeed(requirement);
+    }
+
+    /// <summary>
+    /// Ensure the acting user can delete the requested <see cref="CollectionUser"/> resource(s).
+    /// </summary>
+    /// <remarks>
+    /// Checks that the following are all true:
+    /// - The target collection(s) exists 
+    /// - The acting user is an owner/admin AND the collection management setting is enabled
+    ///   OR
+    ///   The target collection(s) is manageable by the acting user. 
+    /// </remarks>
+    private async Task CanDeleteAsync(AuthorizationHandlerContext context,
+        CollectionUserOperationRequirement requirement, ICollection<CollectionUser> resource)
+    {
+        if (!_currentContext.UserId.HasValue)
+        {
+            context.Fail();
+            return;
+        }
+
+        var distinctTargetCollectionIds = resource.Select(c => c.CollectionId).Distinct().ToList();
+
+        // List of collections the user is performing the operation on
+        var targetCollections = await _collectionRepository.GetManyByManyIdsAsync(distinctTargetCollectionIds);
+
+        // A target collection does not exist, fail the requirement
+        if (targetCollections.Count != distinctTargetCollectionIds.Count)
+        {
+            context.Fail();
+            return;
+        }
+
+        // TODO: Add check for future organization Collection Management setting here and skip the next check if it's enabled
+
+        // List of collections the user is allowed to manage
+        var manageableCollections = (await _collectionRepository.GetManyByUserIdAsync(_currentContext.UserId.Value)).Where(c => c.Manage).ToList();
+
+        // Any target collection is not in the list of manageable collections, fail the requirement
+        if (targetCollections.Any(tc => !manageableCollections.Exists(c => c.Id == tc.Id)))
+        {
+            context.Fail();
+            return;
+        }
+
+        context.Succeed(requirement);
+    }
+
+    /// <summary>
+    /// Ensure the acting user can created the requested <see cref="CollectionUser"/> resource(s) for the
+    /// specified collection (that has not be created in the database yet).
+    /// </summary>
+    /// <remarks>
+    /// Checks that the following are all true:
+    /// - The target collection is provided in the requirement
+    /// - All collection users are assigned to the target collection in the requirement
+    /// - The target users exist and belong to the same organization as the target collection
+    /// </remarks>
+    private async Task CanCreateForNewCollectionAsync(AuthorizationHandlerContext context,
+        CollectionUserOperationRequirement requirement, ICollection<CollectionUser> resource)
```

Comment:
Hmm I see the problem here. Can you push the code where these auth handlers are being used? It'll be helpful to see the complete picture, and also to see the code we're replacing.

I suspect that this Auth Handler includes too much general validation logic. Authorization is just about the user's permissions, not about making sure the general business logic of the request is valid (such as making sure the orgUsers match up to the organization). We might end up moving some of it back out to the controller or command level, which might solve some of our problems here.

Reply:
This was just an initial strategy I wanted to demonstrate and get feedback on after our discussion with @vincentsalucci and his work. The only example for this particular operation requirement is in the unit tests at this time. I'll have a separate PR draft I can put up that demonstrates the `Create` operation shortly.

> I suspect that this Auth Handler includes too much general validation logic.

Totally fair point and I started to get that feeling as well. I felt the "orgUser/organization" check was authorization related because it ensured the acting user was only acting on users it had permission to assign (i.e. belong to the org). 

It sounds like you'd prefer to only check if the target organization users exist and belong to **any** organization the acting user has a role at. If the target user belongs to an unrelated org then the acting user obviously shouldn't have access to them. Then it would be the responsibility of the command/controller to ensure the users and collection match the same organization. 

It think that sounds reasonable and do admit it would likely simplify these authorization handlers quite a bit. I can work on making that change.

Reply:
It's a bit hard to say where to draw the line. Here's how I'm thinking of it (not to be too prescriptive, just to try to be clear):

**Can create collection-user and collection-group associations**
* validation logic in the command: is this a valid request
  * do all these collections exist?
  * do all orgUsers/groups exist?
  * do all orgUsers/groups match the orgId of the collection they're being associated to?
  * maybe we do this check first so that by the time it gets to the handler we can assume it's a valid object
* authorization handler: does the user have permission
  * is the user an admin or owner for that org?
  * or, does the user have Can Manage permissions for that collection?

**Can delete collection-user and collection-group associations**
* validation logic in the command: is this a valid request
  * much less required here, maybe nothing? We can just fire delete requests off to the db
* authorization handler: does the user have permission
  * is the user an admin or owner for that org?
  * or, does the user have Can Manage permissions for that collection?

(I may not have followed this strictly in my own PR, just thinking it through in more detail now)

An advantage of this is that then the `CanCreateForNewCollectionAsync` disappears altogether, it's entirely validation logic. I think if you're creating a collection then it's assumed you can create the initial associations with that collection as part of that same operation.

Reply:
Some null checks may be required as part of the authorization logic (like if the org doesn't exist, we can't check permissions on it). But it's not the main focus.

Reply:
That all sounds reasonable to and makes sense to get validation logic out of the authorization handler.

One downside I see to splitting the validation and authorization logic is that we'll have to query the database in each context. First in the validation logic to see if the entities exist and are configured correctly (everything matches up). Then again when we check if the user has the Can Manage permission for each collection.

Not that this extra database call would be enough reason to avoid the clear separation of concerns, it just seems to be an unfortunate side effect.

Reply:
These changes turned out nicely I think, the handler is more straightforward now and only handles authorization concerns.

Another benefit of the simplification is that the authorization logic for both `CollectionUser` and `CollectionGroup` resources became identical, so I was able to create a shared `CollectionAccessAuthorizationHandlerBase` to use for both resource handlers internally. Each still have their own handlers with different operation requirements for `Create` and `Delete` to follow the authorization/operation pattern and allow for modification for each operation should it be required in the future. For now though, they all call the same underlying `CanManageCollectionAccess` method in the base class to keep things less repetitive and easier to test.

Reply:
I agree this is looking really good! I'll leave a fresh round of feedback below.